File: 2023/1/zig/p1.zig
const std = @import("std");

const MAX_ITERATIONS_COUNT = 100_000;
const CORRECT_ANSWER = 54697;

pub fn main() !void {
    const stdout = std.io.getStdOut().writer();
    const stdin = std.io.getStdIn().reader();

    var buffer: [1024]u8 = undefined;
    var sum: i32 = 0;
    while (try stdin.readUntilDelimiterOrEof(&buffer, '\n')) |line| {
        sum += try getCalibrationValue(line);
    }

    std.debug.assert(sum == CORRECT_ANSWER);
    try stdout.print("\n/2023/d1/P1 result: {d}\n", .{sum});
}

fn getCalibrationValue(s: []const u8) !i32 {
    var start: usize = 0;
    var end: usize = s.len - 1;
    var d1: i32 = -1;
    var d2: i32 = -1;

    var counter: i32 = 0; // guarding from ininite loop
    while (start <= end and counter <= MAX_ITERATIONS_COUNT) {
        if (std.ascii.isDigit(s[start])) {
            if (d1 == -1) {
                d1 = s[start] - '0';
            }
        }
        if (std.ascii.isDigit(s[end])) {
            if (d2 == -1) {
                d2 = s[end] - '0';
            }
        }

        if (d1 == -1) {
            start += 1;
        }
        if (d2 == -1) {
            end -= 1;
        }

        if (d1 != -1 and d2 != -1) {
            break;
        }

        counter += 1;
    }

    if (counter >= MAX_ITERATIONS_COUNT) {
        return error.IterationsCountExhausted;
    }

    if (d1 == -1 or d2 == -1) {
        return error.ParseError;
    }

    return d1 * 10 + d2;
}

File: 2023/1/c/p1.c
#include "p1.h"
#include <assert.h>
#include <ctype.h>
#include <stdio.h>
#include <stdlib.h>

int main(int argc, char **argv) {
  int sum = 0;
  int cv;
  char buffer[1000];
  while (fgets(buffer, sizeof(buffer), stdin) != NULL) {
    if (get_calibration_value((char *)&buffer, &cv) != 0) {
      perror("Error while parsing calibration value\n");
      return 1;
    }
    sum += cv;
  }

  assert(sum == CORRECT_ANSWER);
  printf("\n/2023/d1/P1 result: %d\n", sum);

  return 0;
}

int get_calibration_value(char *s, int *cv) {
  int i;
  short d1 = -1, d2 = -1;

  // iterating over line
  for (i = 0; (s[i] != '\0') && (i <= MAX_ITERATIONS_COUNT); i++) {
    if (!isdigit(s[i])) {
      continue;
    }

    // set the first digit as both d1 and d2
    if (d1 == -1) {
      d1 = s[i] - '0';
      d2 = s[i] - '0';
    } else {
      // last digit overwrites d2
      d2 = s[i] - '0';
    }
  }

  if (i >= MAX_ITERATIONS_COUNT) {
    return 1;
  }

  if (d1 == -1 || d2 == -1) {
    return 1;
  }

  *cv = d1 * 10 + d2;

  return 0;
}
